packages/apps/job-launcher/server/src/modules/job/job-moderation.service.ts

```diff
@@ -0,0 +1,289 @@
+import { ImageAnnotatorClient, protos } from '@google-cloud/vision';
+import { Storage } from '@google-cloud/storage';
+import { VisionConfigService } from '../../common/config/vision-config.service';
+import { ErrorJobModeration } from '../../common/constants/errors';
+import { listObjectsInBucket } from '../../common/utils/storage';
+import { HttpStatus, Injectable, Logger } from '@nestjs/common';
+import {
+  ContentModerationFeature,
+  ContentModerationLevel,
+  JOB_MODERATION_BATCH_SIZE,
+  JOB_MODERATION_MAX_REQUESTS_PER_MINUTE,
+  ONE_MINUTE_IN_MS,
+} from '../../common/constants';
+import {
+  DataModerationResultDto,
+  ImageModerationResultDto,
+  ModerationResultDto,
+} from './job-moderation.dto';
+import { ControlledError } from '../../common/errors/controlled';
+import { StorageService } from '../storage/storage.service';
+import { SlackConfigService } from '../../common/config/slack-config.service';
+import { sendSlackNotification } from '../../common/utils/slack';
+import { JobEntity } from './job.entity';
+import { JobRepository } from './job.repository';
+import { JobStatus } from '../../common/enums/job';
+import { sleep } from '../../common/utils/sleep';
+
+@Injectable()
+export class JobModerationService {
+  private visionClient: ImageAnnotatorClient;
+  public readonly logger = new Logger(JobModerationService.name);
+
+  constructor(
+    private readonly jobRepository: JobRepository,
+    private readonly visionConfigService: VisionConfigService,
+    private readonly slackConfigService: SlackConfigService,
+    private readonly storageService: StorageService,
+  ) {}
+
+  public async jobModeration(jobEntity: JobEntity): Promise<JobEntity> {
+    const manifest = (await this.storageService.downloadJsonLikeData(
+      jobEntity.manifestUrl,
+    )) as any;
+
+    const dataModerationResults: DataModerationResultDto =
+      await this.dataModeration(manifest.data.data_url);
+
+    if (dataModerationResults.containsAbuse) {
+      const abusiveImageLinks = dataModerationResults.veryLikelyOrLikelyResults
+        .map((result) => `- ${result.imageUrl}`)
+        .join('\n');
+
+      const message = `The following images contain abusive content and caused the job to fail:\n${abusiveImageLinks}`;
+
+      this.logger.error(message);
+
+      jobEntity.failedReason = message;
+      jobEntity.status = JobStatus.FAILED;
+      await this.jobRepository.updateOne(jobEntity);
+
+      return jobEntity;
+    }
+
+    // If there are possible moderation issues, save links to GCS and send a Slack notification
+    if (dataModerationResults.possibleResults.length > 0) {
```

Comment:
Just to follow the same line of `positiveAbuseResults`, rename it to `possibleAbuseResults`